beps/0012-metrics-service/README.md

```diff
@@ -0,0 +1,255 @@
+  }
+}
+```
+
+### Naming Conventions
+
+All Backstage metrics MUST follow this hierarchical pattern:
+
+`backstage.{scope}.{scope_name}.{category}.{metric_name}`
+
+**Where:**
+
+- `backstage` is the root namespace for all Backstage metrics
+- `{scope}` is the system scope (`pluginId`, `core` for core services)
+- `{scope_name}` is the name of the scope (e.g. `my-plugin`, `catalog`, `scaffolder`, etc.)
+- `{category}` is the logical grouping within the `{scope}`
+- `{metric_name}` is the name of the metric
+
+The metric name will be the name as provided by the plugin author in the form `{category}.{name}`.
+
+```ts
+const rootMetrics = DefaultRootMetricsService.fromConfig(config);
+const pluginMetricsService = rootMetrics.forPlugin('my-plugin');
+
+const metric = pluginMetricsService.createCounter('my_category.my_metric');
+
+metric.add(1);
+
+// ...
+// metric is now available as `backstage.plugin.my-plugin.my_category.my_metric`
+```
+
+#### Scope
+
+The `scope` represents where it belongs in the Backstage ecosystem.
+
+- `pluginId` - A plugin-specific metric (e.g. `backstage.plugin.catalog.entities.count`)
+- `core` - A core service metric (e.g. `backstage.core.database.connections.active`)
+
+##### Plugin-Scoped Metrics
+
+Pattern: `backstage.plugin.{pluginId}.{category}.{metric_name}`
+
+```yaml
+# Examples
+backstage.plugin.catalog.entities.processed.total
+backstage.plugin.scaffolder.tasks.completed.total
+backstage.plugin.techdocs.builds.active
+backstage.plugin.auth.sessions.active.total # todo: technically a core service and a backend plugin
+```
+
+##### Core Metrics
+
+Pattern: `backstage.core.{core_service}.{category}.{metric_name}`
+
+```yaml
+# Examples
+backstage.core.database.connections.active
+backstage.core.scheduler.tasks.queued.total
+backstage.core.http.requests.total
+```
+
+### References
+
+- [General naming considerations](https://opentelemetry.io/docs/specs/semconv/general/naming/#general-naming-considerations)
+- [Acquiring a meter](https://opentelemetry.io/docs/languages/js/instrumentation/#acquiring-a-meter)
+
+## Release Plan
+
+1. Create a new `RootMetricsService` that initializes metrics and app-wide concerns.
+1. Create the plugin-scoped `MetricsService` that provides a metrics service for plugins.
+1. Create alpha-related documentation to add to existing core service [docs](https://backstage.io/docs/backend-system/core-services/index).
+1. Release the metrics service as alpha.
+1. Refactor catalog and scaffolder implementations to use the new (alpha) `MetricsService`.
+1. Release the metrics service
+1. Offer a migration path for existing adopters to migrate to the new metrics service.
+1. Update all documentation to reference the new metrics service.
+1. Create follow-up action items to integrate the new metrics service into the core system.
+1. Fully deprecate all existing metrics implementations like the existing Prometheus one-off implementations.
```

Comment:
Should we do this sooner, like now? At least for things in the main Backstage repo.

Reply:
Yeah, it's worth marking them deprecated as one of the first tasks in this list.